go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,460 @@
+/*
+Copyright 2019 The Vitess Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysqlctl
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/klauspost/pgzip"
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/mysqlctl/backupstorage"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// XtrabackupEngine encapsulates the logic of the xtrabackup engine
+// it implements the BackupEngine interface and contains all the logic
+// required to implement a backup/restore by invoking xtrabackup with
+// the appropriate parameters
+type XtrabackupEngine struct {
+}
+
+var (
+	// path where backup engine program is located
+	xtrabackupEnginePath = flag.String("xtrabackup_root_path", "", "directory location of the xtrabackup executable, e.g., /usr/bin")
+	// flags to pass through to backup engine
+	xtrabackupBackupFlags = flag.String("xtrabackup_backup_flags", "", "flags to pass to backup command. these will be added to the end of the command")
+	// flags to pass through to restore phase
+	xbstreamRestoreFlags = flag.String("xbstream_restore_flags", "", "flags to pass to xbstream command during restore. these will be added to the end of the command. These need to match the ones used for backup e.g. --compress / --decompress, --encrypt / --decrypt")
+	// streaming mode
+	xtrabackupStreamMode = flag.String("xtrabackup_stream_mode", "tar", "which mode to use if streaming, valid values are tar and xbstream")
+	xtrabackupUser       = flag.String("xtrabackup_user", "", "User to use for xtrabackup")
```

Comment:
nit - I would make clear in the comment that this is the user for mysql that xtrabackup will use.

Reply:
How does this look? 
```User that xtrabackup will use to connect to the database server. This user must have all necessary privileges. For details, please refer to xtrabackup documentation.```